packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,326 @@
+    <nimble-table-column-icon data-key="ready"></nimble-table-column-icon>
+    ...
+</nimble-table>
+```
+
+These columns will _not_ have templates/CSS associated with them. Instead, each column element will provide a FAST ViewTemplate for the visual representation of each cell in the column. The ordering of the columns in the markup will determine the visual ordering of the columns (top to bottom equals left to right...unless in 'rtl'). Re-ordering of columns will be done, at least at first, through the re-ordering of the column elements in the DOM.
+
+The table API to support this could look like the following:
+
+```TS
+public Table<TableRecord> extends FoundationElement {
+    ...
+
+    /*
+     * @internal
+     */
+    @observable
+    public readonly slottedColumns: TableColumn[] = [];
+
+    private slottedColumnsChanged(): void {
+        if (this.slottedColumns.length > 0) {
+            for (const tableColumn of this.slottedColumns) {
+                if (this.isTableColumn(tableColumn)) {
+                    // do init work
+                }
+            }
+        }
+    }
+}
+
+template:
+<template>
+    <div>
+        ...
+    </div>
+    <slot ${slotted('slottedColumns')}></slot>
+</template>
+```
+
+### `TableCellData` interface
+
+A table cell represents a single column for a single row. The data that a cell has access to will be a subset of the data for the entire row. An instance of a table cell will be generic to describe the subset of data it contains, where the `TCellData` type is a superset of the type represented by [`TableRecord`](https://github.com/ni/nimble/blob/3e4b8d3dd59431d1671e381aa66052db57bc475c/packages/nimble-components/src/table/types.ts#L24):
+
+```TS
+interface TableCellState<TCellData extends TableRecord> {
+  data: TCellData;
+}
+```
+
+This interface could possibly be expanded in the future to communicate relevant table state to the cell template, such as whether or not the row is selected.
+
+### `TableColumn<>`
+
+This abstract class is what a column web component (i.e. a slotted column) must extend.
+
+```TS
+abstract class TableColumn<TCellData extends TableRecord = TableRecord> {
+    // This method will produce the expected TableCellState that the cellTemplate expects as
+    // its source
+    abstract generateCellState(cellData: TCellData): TableCellState<TCellData>;
+
+    // The template to use to render the cell content for the column
+    abstract cellTemplate: ViewTemplate<TableCellState<TCellData>>;
+
+    // The style to apply to the cellTemplate
+    cellStyles?: ElementStyles;
+
+    // The keys that should be present in TCellData.
+    // This array is parallel with the keys returned from `getDataKeys()`.
+    readonly cellDataKeyNames: readonly string[];
+
+    // The keys from the row data that correlate to the data that will be in TCellData.
+    // This array is parallel with the keys specified by `dataKeyNames`.
+    abstract getRowDataKeys(): string[];
```

Comment:
both this and the above result in keynames right?

```suggestion
    abstract getRowDataKeyNames(): string[];
```

Reply:
That is correct. I think these names should actually be updated with the new terminology, so I think something like this would be more appropriate:
```
// The field names that should be present in the TCellData record.
// This array is parallel with the field names returned from `getRowRecordFieldNames()`.
readonly cellRecordFieldNames: readonly TableFieldName[];

// The names of the fields in the row's record that will be used to populate the TCellData record.
// This array is parallel with the keys specified by `cellRecordFieldNames`.
abstract getRowRecordFieldNames(): TableFieldName[];
```

Reply:
What about cellStateFieldNames and getRecordFieldNames?

Thinking there really isn't a cell record, just cell state which has pieces of a record.

Reply:
That seems reasonable to me. If we want to get really specific, maybe the property should be cellStateDataFieldNames since they aren't fields within the cell state itself but fields within the data in the cell state. I'm undecided on whether or not it's worth making that distinction. One reason it might be worth making the distinction is if we ever think we'll need an API for getting the field names that a column adds to the cell state. I can't think of a reason why we'd need to do that though.

Reply:
Does this imply that the type parameter should change to `TCellRecord`, and that then the `TableCellState` interface should have a member called `record` and not `data`? My feeling is that this doesn't seem necessary, but I can be convinced otherwise.

Reply:
No, I don't think this implies that the type parameter or member need to be renamed.

Reply:
Updated.